shared/src/main/scala/tastyquery/ast/TypeTrees.scala

```diff
@@ -35,45 +37,95 @@ object TypeTrees {
+        }
+      )
+  }
 
   /** qualifier#name */
-  case class SelectTypeTree(qualifier: TypeTree, name: TypeName) extends TypeTree
+  case class SelectTypeTree(qualifier: TypeTree, name: TypeName) extends TypeTree {
+    override protected def calculateType: Type =
+      TypeRef(qualifier.toType, name)
+  }
 
-  /** qualifier.TypeName */
-  case class SelectTypeFromTerm(qualifier: Tree, name: TypeName) extends TypeTree
+  /** qualifier.name */
+  case class TermRefTypeTree(qualifier: Tree, name: TermName) extends TypeTree {
+    override protected def calculateType: Type =
+      TermRef(qualifier.tpe, name)
+  }
 
   /** arg @annot */
-  case class AnnotatedTypeTree(tpt: TypeTree, annotation: Tree) extends TypeTree
+  case class AnnotatedTypeTree(tpt: TypeTree, annotation: Tree) extends TypeTree {
+    override protected def calculateType: Type =
+      AnnotatedType(tpt.toType, annotation)
+  }
 
   /** [bound] selector match { cases } */
-  case class MatchTypeTree(bound: TypeTree, selector: TypeTree, cases: List[TypeCaseDef]) extends TypeTree
+  case class MatchTypeTree(bound: TypeTree, selector: TypeTree, cases: List[TypeCaseDef]) extends TypeTree {
+    override protected def calculateType: Type =
+      throw new TypeTreeToTypeError(this) // TODO
+  }
 
   case class TypeCaseDef(pattern: TypeTree, body: TypeTree)
 
   case class TypeTreeBind(name: TypeName, body: TypeTree, override val symbol: RegularSymbol)
       extends TypeTree
-      with DefTree(symbol)
+      with DefTree(symbol) {
+    override protected def calculateType: Type =
+      TermRef(NoType, symbol)
+  }
 
-  case object EmptyTypeTree extends TypeTree
+  case object EmptyTypeTree extends TypeTree {
+    override protected def calculateType: Type =
+      NoType
+  }
 
-  case class TypeBoundsTree(low: TypeTree, high: TypeTree)
+  case class TypeBoundsTree(low: TypeTree, high: TypeTree) {
+    def toTypeBounds: TypeBounds =
+      RealTypeBounds(low.toType, high.toType)
+  }
 
   /** >: lo <: hi
     *  >: lo <: hi = alias  for RHS of bounded opaque type
     */
-  case class BoundedTypeTree(bounds: TypeBoundsTree, alias: TypeTree) extends TypeTree
-  case class NamedTypeBoundsTree(name: TypeName, bounds: TypeBounds) extends TypeTree
+  case class BoundedTypeTree(bounds: TypeBoundsTree, alias: TypeTree) extends TypeTree {
+    override protected def calculateType: Type =
+      BoundedType(bounds.toTypeBounds, alias.toType)
+  }
+
+  case class NamedTypeBoundsTree(name: TypeName, bounds: TypeBounds) extends TypeTree {
+    override protected def calculateType: Type =
+      NamedTypeBounds(name, bounds)
+  }
 
-  case class TypeLambdaTree(tparams: List[TypeParam], body: TypeTree) extends TypeTree
+  case class TypeLambdaTree(tparams: List[TypeParam], body: TypeTree) extends TypeTree {
+    override protected def calculateType: Type =
+      TypeLambda(tparams)(binders => body.toType) // TODO Shouldn't we use the binders here?
```

Comment:
the binders parameter is just a hack for modifying the context when we see `TYPELAMBDAtype` - it can be ignored in favour of the `tparams` here.
Here is what dotty does:
```scala
def assignType(tree: untpd.LambdaTypeTree, tparamDefs: List[TypeDef], body: Tree)(using Context): LambdaTypeTree =
    tree.withType(HKTypeLambda.fromParams(tparamDefs.map(_.symbol.asType), body.tpe))
```